### DISCUSSION TITLE
Any ideas for pre-parsing a JSON file for replacement of environment variables?

### DISCUSSION THREAD

**Discussion by Tails86** [nlohmann/json#4501] [Q&A] [answered]
I'm looking for ideas about how to pre-parse a json file which contains variable callouts in the format `${VAR}`. Currently, I am using the callback in order replace string key/value items with a new string. This mostly works for me except for a couple of issues:
1. I can't think of a good way to escape a $ if the variable shouldn't be parsed - the JSON parsing throws in exception because `"\$"` is an invalid escapement, and it wouldn't matter if it didn't because it would have escaped it already before getting to this point anyway.
2. There is no great way to parse number or boolean values this way - I thought about maybe determining if the string should be converted within the callback, but that seems cumbersome.

What I think would be nice is if I could pre-parse the file line-by-line and feed it into the json parser, but there doesn't seem to be a way to do that. I'd have to read the entire file into a modified string, stringstream, or temp file and then feed it to the json parser.

Thanks for your time

**Comment by gregmarr** [comment, marked as answer]
Sounds like you want an input adapter.  https://github.com/nlohmann/json/blob/develop/include/nlohmann/detail/input/input_adapters.hpp
`wide_string_input_adapter` is an example of converting text between when it's read from the underlying adapter and when it's provided to the parser.

**Reply by Tails86** [reply]
Huh, yeah. That's looks exactly like what I need. Thanks a lot!

**Reply by gregmarr** [reply]
You're welcome.

**Reply by gregmarr** [reply]
@Tails86 If this resolved your question, can you mark it as answered?
